config/reflection_magic.go

```diff
@@ -121,18 +143,28 @@ func makeConstructor(
 	if err != nil {
 		return nil, err
 	}
+	if err := checkConstructorOpts(t, opts...); err != nil {
+		return nil, err
+	}
 
 	return func(h host.Host, u *tptu.Upgrader, cg connmgr.ConnectionGater) (interface{}, error) {
-		arguments := make([]reflect.Value, len(argConstructors))
+		arguments := make([]reflect.Value, 0, len(argConstructors)+len(opts))
 		for i, makeArg := range argConstructors {
 			if arg := makeArg(h, u, cg); arg != nil {
-				arguments[i] = reflect.ValueOf(arg)
+				arguments = append(arguments, reflect.ValueOf(arg))
 			} else {
 				// ValueOf an un-typed nil yields a zero reflect
 				// value. However, we _want_ the zero value of
 				// the _type_.
-				arguments[i] = reflect.Zero(t.In(i))
+				arguments = append(arguments, reflect.Zero(t.In(i)))
+			}
+		}
+		for _, opt := range opts {
+			// don't panic on nil options
+			if opt == nil {
+				return nil, errors.New("expected a transport option, got nil")
```

Comment:
did you not reject untyped nil options further back already?

Reply:
Good catch, I removed the redundant check.